include/libhal-expander/tla2528.hpp

```diff
@@ -0,0 +1,178 @@
+#pragma once
+#include <libhal/i2c.hpp>
+#include <libhal/steady_clock.hpp>
+#include <libhal/units.hpp>
+
+namespace hal::expander {
+
+// adapters
+class tla2528_adc;
+class tla2528_input_pin;
+class tla2528_output_pin;
+
+/**
+ * @brief tla2528 is a gpio expander & adc mux driver
+ *
+ * tla2528 has 8 pins which can be independently operated as an adc,
+ * digital input, and digital out over i2c. The i2c address is configured by
+ * resistors connected to the chip There are no options for internal pull up or
+ * pull down resistors. The output pins have the option of push-pull or
+ * open-drain. When in adc mode there is an option (UNIMPLEMENTED) to increase
+ * reading granularity though sampling averaging.
+ */
+class tla2528
+{
+public:
+  enum class pin_mode : hal::byte
+  {
+    analog_input,
+    digital_input,
+    digital_output_open_drain,
+    digital_output_push_pull
+  };
+
+  /**
+   * @param p_i2c i2c bus of the device
+   *
+   * @param p_i2c_address i2c address configured on the tla, by default is set
+   * to the i2c address of no resistors attached to address config pins.
+   */
+  tla2528(hal::i2c& p_i2c, hal::byte p_i2c_address = default_address);
+
+  /**
+   * @brief set what service a pin will provide
+   *
+   * @param p_mode tla2528::pin_mode enum of desired pin mode
+   *
+   * @param p_channel which pin to configure
+   *
+   * @throws hal::argument_out_of_domain - if p_channel out of range (>7)
+   *
+   * @throws hal::resource_unavailable_try_again - if adapters are made for a
+   * pin an exception may be thrown to prevent invalid behavior
+   *
+   * @throws hal::no_such_device - no device responded on i2c bus
+   */
+  void set_pin_mode(pin_mode p_mode, hal::byte p_channel);
+
+  /**
+   * @brief set digital output level pin
+   *
+   * @param p_channel pin to set output
+   *
+   * @param p_high the output level of the pin, true is high, false is low.
+   *
+   * @throws hal::argument_out_of_domain - if p_channel out of range (>7)
+   *
+   * @throws hal::no_such_device - no device responded on i2c bus
+   *
+   * The device will write to a register that caches the desired output state.
+   * When a pin is in a digital output mode it will reference the desired state
+   * cache. If the pin is not set to digital output it will just be stored.
+   */
```

Comment:
```suggestion
  /**
   * @brief set digital output level pin
   *
   * The device will write to a register that caches the desired output state.
   * When a pin is in a digital output mode it will reference the desired state
   * cache. If the pin is not set to digital output it will just be stored.
   *
   * @param p_channel pin to set output
   * @param p_high the output level of the pin, true is high, false is low.
   * @throws hal::argument_out_of_domain - if p_channel out of range (>7)
   * @throws hal::no_such_device - no device responded on i2c bus
   */
```

Put the description after the brief then define the parameters and throwables. But I also don't think the description makes sense to a reader. I think the underlying implementation, in this case, isn't important enough to spell out to the API user. In other cases I could see a rationale for stating the implementation details in order to warn a user about a potentially expensive operation. But in this case, I don't see it as very useful.

Reply:
Should I remove the descriptions, improve the clarity, or was it just unnecessary work that can be left alone?

Either way, I changed the order as requested.

Reply:
Reading it over again, you can probably eliminate that section. If anything may be useful to be as a comment in the source file for readers looking to understand the code. I'll leave it to you if you want to place this somewhere in the source.